File: compiler/nodes/expressions/is.go
package expressions

import (
	"wipple/codegen"
	"wipple/database"
	"wipple/nodes/patterns"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type IsExpressionNode struct {
	Left  database.Node
	Right database.Node
	Facts *database.Facts

	inputTemporary database.Node
	trueVariant    database.Node
	falseVariant   database.Node
}

func (node *IsExpressionNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseIsExpression(parser *syntax.Parser) (*IsExpressionNode, *syntax.Error) {
	span := parser.Spanned()

	left, err := ParseExpressionElement(parser)
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("IsOperator")
	if err != nil {
		return nil, err
	}

	parser.ConsumeLineBreaks()

	right, err := patterns.ParsePatternElement(parser)
	if err != nil {
		return nil, err
	}

	return &IsExpressionNode{
		Left:  left,
		Right: right,
		Facts: database.NewFacts(span()),
	}, nil
}

func (node *IsExpressionNode) Visit(visitor *visit.Visitor) {
	visitExpression(visitor, node)

	visitor.Visit(node.Left)

	node.inputTemporary = &database.HiddenNode{Facts: database.NewFacts(database.GetSpanFact(node.Left))}
	visitor.Db.Register(node.inputTemporary)
	visit.Matching(visitor, node.inputTemporary, false, func() struct{} {
		visitor.Visit(node.Right)
		return struct{}{}
	})

	visitor.Constraint(typecheck.GroupConstraint(node.inputTemporary, node.Left))

	booleanTypeDefinition, ok := visit.Resolve[*visit.TypeDefinition](visitor, "Boolean", node)
	if !ok {
		return
	}

	trueVariant, ok := visit.Resolve[*visit.VariantConstructorDefinition](visitor, "True", node)
	if !ok {
		return
	}

	falseVariant, ok := visit.Resolve[*visit.VariantConstructorDefinition](visitor, "False", node)
	if !ok {
		return
	}

	node.trueVariant = trueVariant.Node
	node.falseVariant = falseVariant.Node

	visitor.Constraint(typecheck.GroupConstraint(node, booleanTypeDefinition.Node))
}

func (node *IsExpressionNode) Codegen(c *codegen.Codegen) error {
	if node.inputTemporary == nil || node.trueVariant == nil || node.falseVariant == nil {
		return c.Error(node)
	}

	span := database.GetSpanFact(node)

	whenExpression := &WhenExpressionNode{
		Input: node.Left,
		Arms: []Arm{
			{
				Pattern: node.Right,
				Value:   node.trueVariant,
				Span:    span,
			},
			{
				Pattern: &patterns.WildcardPatternNode{
					Facts: database.NewFacts(span),
				},
				Value: node.falseVariant,
				Span:  span,
			},
		},
		Facts:          database.NewFacts(span),
		inputTemporary: node.inputTemporary,
	}

	return c.Write(whenExpression)
}

File: compiler/src/nodes/expressions/is.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenResult, ir},
    database::{Fact, HiddenNode, Node, NodeRef, Render},
    nodes::{
        ConstructorExpressionNode, parse_expression_element, parse_pattern_element,
        visit_expression,
    },
    syntax::{ParseError, Parser, TokenKind},
    typecheck::GroupConstraint,
    visit::{Visit, Visitor},
};

#[derive(Debug, Clone)]
pub struct ResolvedIs {
    pub true_variant: NodeRef,
    pub false_variant: NodeRef,
}

impl Fact for ResolvedIs {}

impl Render for ResolvedIs {}

#[derive(Debug)]
pub struct IsExpressionNode {
    pub left: NodeRef,
    pub right: NodeRef,
}

impl Node for IsExpressionNode {}

pub fn parse_is_expression(parser: &mut Parser<'_>) -> Result<IsExpressionNode, ParseError> {
    let left = parse_expression_element(parser)?;
    parser.token(TokenKind::IsOperator)?;
    parser.consume_line_breaks();
    let right = parse_pattern_element(parser)?;

    Ok(IsExpressionNode { left, right })
}

impl Visit for IsExpressionNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_expression(node, visitor);

        visitor.visit(&self.left);
        visitor.edge(&self.left, node, "left");

        let span = visitor.span(node);

        visitor.matching(&self.left, true, false, |visitor| {
            visitor.current_match().root = Some(self.left.clone());
            visitor.visit(&self.right);
            visitor.edge(&self.right, node, "right");
        });

        let true_node = visitor.node(
            span.clone(),
            HiddenNode::new(ConstructorExpressionNode {
                constructor: String::from("True"),
            }),
        );

        let false_node = visitor.node(
            span,
            HiddenNode::new(ConstructorExpressionNode {
                constructor: String::from("False"),
            }),
        );

        visitor.visit(&true_node);
        visitor.visit(&false_node);

        visitor.constraint(GroupConstraint::new(node.clone(), true_node.clone()));
        visitor.constraint(GroupConstraint::new(node.clone(), false_node.clone()));

        visitor.insert(
            node,
            ResolvedIs {
                true_variant: true_node,
                false_variant: false_node,
            },
        );
    }
}

impl Codegen for IsExpressionNode {
    fn codegen(&self, node: &NodeRef, ctx: &mut CodegenCtx<'_>) -> CodegenResult {
        let ResolvedIs {
            true_variant,
            false_variant,
        } = ctx
            .get(node)
            .ok_or_else(|| anyhow::format_err!("unresolved"))?;

        ctx.codegen(&self.left)?;

        ctx.push_conditions();
        ctx.codegen(&self.right)?;
        let conditions = ctx.pop_conditions();

        ctx.push_instructions();
        ctx.codegen(&true_variant)?;
        let true_instructions = ctx.pop_instructions();

        ctx.push_instructions();
        ctx.codegen(&false_variant)?;
        let false_instructions = ctx.pop_instructions();

        ctx.instruction(ir::Instruction::If {
            node: Some(node.clone()),
            branches: vec![(conditions, true_instructions, Some(true_variant.clone()))],
            else_branch: Some((false_instructions, Some(false_variant.clone()))),
        });

        Ok(())
    }
}
